index.js

```diff
@@ -1171,14 +1164,23 @@
                       var t = expType.types[k].type;
                       return Z.chain(function(r) {
                         return test(env, t, x) ? Right(r) : Left(function() {
-                          return invalidValue(env,
-                                              typeInfo,
-                                              index,
-                                              Z.concat(propPath, [k]),
-                                              x);
+                          var propPath$ = Z.concat(propPath, [k]);
+                          return t.type === VARIABLE ?
+                            (function($valuesByPath) {
+                              var k = JSON.stringify(Z.concat([index],
+                                                              propPath$));
+                              var valuesByPath = {};
+                              valuesByPath[k] = [x];
+                              return typeVarConstraintViolation(env,
+                                                                typeInfo,
+                                                                index,
+                                                                propPath$,
+                                                                valuesByPath);
+                            }()) :
+                            invalidValue(env, typeInfo, index, propPath$, x);
                         });
                       }, e);
-                    }, Right(r), innerValues)
+                    }, e, innerValues)
```

Comment:
This was a type error that went undetected for a long time. `r` was in fact a value of type `Either a b`, so `Right(r)` produced a value of type `Either a (Either a b)` which was incorrect. The correct expression was simply `r`.

I renamed the parameter to hint that it's of type `Either a b`.